File: packages/py-moose-lib/moose_lib/dmv2/materialized_view.py
"""
Materialized View definitions for Moose Data Model v2 (dmv2).

This module provides classes for defining Materialized Views,
including their SQL statements, target tables, and dependencies.

Two types of materialized views are supported:
- **MaterializedView**: Incremental MV triggered on every insert to source tables
- **RefreshableMaterializedView**: Scheduled MV that runs on a refresh interval
"""

from dataclasses import dataclass, field
from typing import Any, Literal, Optional, Union, Generic, List

from pydantic import BaseModel, ConfigDict

from ..blocks import ClickHouseEngines, MergeTreeEngine
from .types import BaseTypedResource, T
from .olap_table import OlapTable, OlapConfig
from .life_cycle import LifeCycle
from ._registry import _materialized_views
from ._source_capture import get_source_file_from_stack
from .view import View


def _format_table_reference(table: Union[OlapTable, View]) -> str:
    """Helper function to format a table reference as `database`.`table` or just `table`"""
    database = table.config.database if isinstance(table, OlapTable) else None
    if database:
        return f"`{database}`.`{table.name}`"
    return f"`{table.name}`"


# ============================================================================
# Refresh Configuration Types for Refreshable Materialized Views
# ============================================================================

# Supported time units for refresh intervals
TimeUnit = Literal["second", "minute", "hour", "day", "week", "month", "year"]


@dataclass
class RefreshIntervalEvery:
    """
    Refresh interval using EVERY mode - periodic refresh at fixed times.
    Example: RefreshIntervalEvery(1, "hour") => REFRESH EVERY 1 HOUR
    """

    value: int
    """The numeric value of the interval"""
    unit: TimeUnit
    """The time unit for the interval"""
    type: Literal["every"] = "every"

    def __post_init__(self):
        if self.value <= 0:
            raise ValueError(
                f"Refresh interval value must be positive, got {self.value}"
            )


@dataclass
class RefreshIntervalAfter:
    """
    Refresh interval using AFTER mode - refresh after interval since last refresh.
    Example: RefreshIntervalAfter(30, "minute") => REFRESH AFTER 30 MINUTE
    """

    value: int
    """The numeric value of the interval"""
    unit: TimeUnit
    """The time unit for the interval"""
    type: Literal["after"] = "after"

    def __post_init__(self):
        if self.value <= 0:
            raise ValueError(
                f"Refresh interval value must be positive, got {self.value}"
            )


RefreshInterval = Union[RefreshIntervalEvery, RefreshIntervalAfter]


@dataclass
class Duration:
    """
    A duration specified as value + unit.
    Used for offset and randomize configurations.
    """

    value: int
    """The numeric value"""
    unit: TimeUnit
    """The time unit"""

    def __post_init__(self):
        if self.value <= 0:
            raise ValueError(f"Duration value must be positive, got {self.value}")


@dataclass
class RefreshConfig:
    """
    Configuration for refreshable (scheduled) materialized views.

    Refreshable MVs run on a schedule (REFRESH EVERY/AFTER) rather than
    being triggered by inserts to source tables.

    Example:
        >>> config = RefreshConfig(
        ...     interval=RefreshIntervalEvery(1, "hour"),
        ...     offset=Duration(5, "minute"),
        ...     depends_on=[hourly_stats_mv],  # Type-safe: only accepts RefreshableMaterializedView
        ... )
    """

    interval: RefreshInterval
    """The refresh interval (EVERY or AFTER)"""
    offset: Optional[Duration] = None
    """Optional offset from interval start. NOTE: Only valid with REFRESH EVERY, not REFRESH AFTER."""
    randomize: Optional[Duration] = None
    """Optional randomization window"""
    depends_on: List["RefreshableMaterializedView"] = field(default_factory=list)
    """Other refreshable MVs this one depends on. Only accepts RefreshableMaterializedView objects."""
    append: bool = False
    """Use APPEND mode instead of full refresh"""


@dataclass
class ResolvedRefreshConfig:
    """
    Internal representation of RefreshConfig with depends_on resolved to string names.
    """

    interval: RefreshInterval
    """The refresh interval (EVERY or AFTER)"""
    offset: Optional[Duration] = None
    """Optional offset from interval start"""
    randomize: Optional[Duration] = None
    """Optional randomization window"""
    depends_on: List[str] = field(default_factory=list)
    """Names of other MVs this one depends on (resolved to strings)"""
    append: bool = False
    """Use APPEND mode instead of full refresh"""


# ============================================================================
# Incremental MaterializedView Configuration and Class
# ============================================================================


class MaterializedViewOptions(BaseModel):
    """Configuration options for creating an incremental Materialized View.

    Incremental MVs are triggered on every insert to source tables.

    Attributes:
        select_statement: The SQL SELECT statement defining the view's data.
        select_tables: List of source tables/views the select statement reads from.
        table_name: Optional name of the underlying target table storing the materialized data.
        materialized_view_name: The name of the MATERIALIZED VIEW object itself.
        engine: Optional ClickHouse engine for the target table.
        order_by_fields: Optional ordering key for the target table.
        metadata: Optional metadata dictionary.
        model_config: ConfigDict for Pydantic validation
        life_cycle: Optional lifecycle management policy. Controls how Moose handles
                    this materialized view when code definitions change. Valid values:
                    LifeCycle.FULLY_MANAGED (default) — Moose auto-creates, updates
                    (via DROP+CREATE), and drops the MV; LifeCycle.DELETION_PROTECTED —
                    Moose auto-creates but will not drop or update the MV;
                    LifeCycle.EXTERNALLY_MANAGED — Moose will not create, update, or
                    drop the MV. Defaults to FULLY_MANAGED when not specified.
    """

    select_statement: str
    select_tables: List[Union[OlapTable, "View"]]
    table_name: Optional[str] = None
    materialized_view_name: str
    engine: Optional[ClickHouseEngines] = None
    order_by_fields: Optional[List[str]] = None
    metadata: Optional[dict] = None
    life_cycle: Optional[LifeCycle] = None
    model_config = ConfigDict(arbitrary_types_allowed=True)


class MaterializedView(BaseTypedResource, Generic[T]):
    """Represents an incremental Materialized View in ClickHouse.

    Incremental MVs are triggered on every insert to source tables.
    For scheduled/refreshable MVs, use `RefreshableMaterializedView` instead.

    Args:
        options: Configuration defining the select statement, names, and dependencies.
        target_table: Optional existing OlapTable to use as the target.
        t: The Pydantic model defining the schema of the target table.

    Attributes:
        target_table (OlapTable[T]): The `OlapTable` instance storing the materialized data.
        config (MaterializedViewOptions): The configuration options used to create the view.
        name (str): The name of the MATERIALIZED VIEW object.
        select_sql (str): The SELECT SQL statement.
        source_tables (list[str]): Names of source tables the SELECT reads from.
        life_cycle (LifeCycle | None): Lifecycle management policy. Controls how Moose
            handles this MV when code definitions change. Defaults to FULLY_MANAGED when
            not specified. See LifeCycle enum for available values.
    """

    kind: str = "MaterializedView"
    target_table: OlapTable[T]
    config: MaterializedViewOptions
    name: str
    select_sql: str
    source_tables: List[str]
    metadata: Optional[dict] = None
    refresh_config: None = None  # Always None for incremental MVs
    life_cycle: Optional[LifeCycle] = None

    def __init__(
        self,
        options: MaterializedViewOptions,
        target_table: Optional[OlapTable[T]] = None,
        **kwargs: Any,
    ):
        self._set_type(options.materialized_view_name, self._get_type(kwargs))

        # Resolve target table
        if target_table:
            self.target_table = target_table
            if self._t != target_table._t:
                raise ValueError(
                    "Target table must have the same type as the materialized view"
                )
        else:
            if not options.table_name:
                raise ValueError(
                    "Name of target table is not specified. "
                    "Provide 'target_table' parameter or 'table_name' in options."
                )
            target_table = OlapTable(
                name=options.table_name,
                config=OlapConfig(
                    order_by_fields=options.order_by_fields or [],
                    engine=options.engine,
                ),
                t=self._t,
            )
            self.target_table = target_table

        if target_table.name == options.materialized_view_name:
            raise ValueError(
                "Target table name cannot be the same as the materialized view name"
            )

        self.name = options.materialized_view_name
        self.config = options
        self.select_sql = options.select_statement
        self.source_tables = [_format_table_reference(t) for t in options.select_tables]
        self.refresh_config = None
        self.life_cycle = options.life_cycle

        # Initialize metadata
        if options.metadata:
            self.metadata = (
                options.metadata.copy()
                if isinstance(options.metadata, dict)
                else options.metadata
            )
        else:
            self.metadata = {}

        # Capture source file from stack trace
        if not isinstance(self.metadata, dict):
            self.metadata = {}
        if "source" not in self.metadata:
            source_file = get_source_file_from_stack()
            if source_file:
                self.metadata["source"] = {"file": source_file}

        if self.name in _materialized_views:
            raise ValueError(f"MaterializedView with name {self.name} already exists")
        _materialized_views[self.name] = self

    def is_incremental(self) -> bool:
        """Returns True - incremental MVs are always incremental."""
        return True

    def is_refreshable(self) -> bool:
        """Returns False - incremental MVs are never refreshable."""
        return False


# ============================================================================
# Refreshable MaterializedView Configuration and Class
# ============================================================================


class RefreshableMaterializedViewOptions(BaseModel):
    """Configuration options for creating a Refreshable Materialized View.

    Refreshable MVs run on a schedule (REFRESH EVERY/AFTER) rather than
    being triggered by inserts to source tables.

    Note: Refreshable MVs always use MergeTree engine (no custom engine option).

    Attributes:
        select_statement: The SQL SELECT statement defining the view's data.
        select_tables: List of source tables/views the select statement reads from.
        materialized_view_name: The name of the MATERIALIZED VIEW object itself.
        target_table_name: Name of the underlying target table.
        order_by_fields: Optional ordering key for the target table.
        refresh_config: Configuration for the refresh schedule. Required.
        metadata: Optional metadata dictionary.
    """

    select_statement: str
    select_tables: List[Union[OlapTable, "View"]]
    materialized_view_name: str
    target_table_name: str
    order_by_fields: Optional[List[str]] = None
    refresh_config: RefreshConfig
    metadata: Optional[dict] = None
    model_config = ConfigDict(arbitrary_types_allowed=True)


class RefreshableMaterializedView(BaseTypedResource, Generic[T]):
    """Represents a Refreshable Materialized View in ClickHouse.

    Refreshable MVs run on a schedule (REFRESH EVERY/AFTER) rather than
    being triggered by inserts to source tables.

    For incremental/trigger-based MVs, use `MaterializedView` instead.

    Args:
        options: Configuration defining the select statement, names, and dependencies.
        t: The Pydantic model defining the schema of the target table.

    Attributes:
        target_table (OlapTable[T]): The `OlapTable` instance storing the materialized data.
        config (RefreshableMaterializedViewOptions): The configuration options.
        name (str): The name of the MATERIALIZED VIEW object.
        select_sql (str): The SELECT SQL statement.
        source_tables (list[str]): Names of source tables the SELECT reads from.
        refresh_config (ResolvedRefreshConfig): The refresh configuration with resolved dependencies.
    """

    kind: str = "MaterializedView"
    target_table: OlapTable[T]
    config: RefreshableMaterializedViewOptions
    name: str
    select_sql: str
    source_tables: List[str]
    metadata: Optional[dict] = None
    refresh_config: ResolvedRefreshConfig  # Always set for refreshable MVs

    def __init__(
        self,
        options: RefreshableMaterializedViewOptions,
        **kwargs: Any,
    ):
        self._set_type(options.materialized_view_name, self._get_type(kwargs))

        # Create target table (always MergeTree for refreshable MVs)
        target_table = OlapTable(
            name=options.target_table_name,
            config=OlapConfig(
                order_by_fields=options.order_by_fields or [],
                engine=MergeTreeEngine(),
            ),
            t=self._t,
        )
        self.target_table = target_table

        if target_table.name == options.materialized_view_name:
            raise ValueError(
                "Target table name cannot be the same as the materialized view name"
            )

        # Validate OFFSET is not used with REFRESH AFTER (only valid with REFRESH EVERY)
        if (
            options.refresh_config.interval.type == "after"
            and options.refresh_config.offset is not None
        ):
            raise ValueError(
                "OFFSET is only valid with REFRESH EVERY, not REFRESH AFTER. "
                "Remove the 'offset' option or change the interval type to 'every'."
            )

        self.name = options.materialized_view_name
        self.config = options
        self.select_sql = options.select_statement
        self.source_tables = [_format_table_reference(t) for t in options.select_tables]

        # Convert refresh_config.depends_on from RefreshableMaterializedView objects to string names
        dep_names = [dep.name for dep in options.refresh_config.depends_on]
        if options.materialized_view_name in dep_names:
            raise ValueError(
                f"RefreshableMaterializedView '{options.materialized_view_name}' "
                "cannot depend on itself"
            )
        self.refresh_config = ResolvedRefreshConfig(
            interval=options.refresh_config.interval,
            offset=options.refresh_config.offset,
            randomize=options.refresh_config.randomize,
            depends_on=dep_names,
            append=options.refresh_config.append,
        )

        # Initialize metadata
        if options.metadata:
            self.metadata = (
                options.metadata.copy()
                if isinstance(options.metadata, dict)
                else options.metadata
            )
        else:
            self.metadata = {}

        # Capture source file from stack trace
        if not isinstance(self.metadata, dict):
            self.metadata = {}
        if "source" not in self.metadata:
            source_file = get_source_file_from_stack()
            if source_file:
                self.metadata["source"] = {"file": source_file}

        if self.name in _materialized_views:
            raise ValueError(
                f"RefreshableMaterializedView with name {self.name} already exists"
            )
        _materialized_views[self.name] = self

    def is_incremental(self) -> bool:
        """Returns False - refreshable MVs are never incremental."""
        return False

    def is_refreshable(self) -> bool:
        """Returns True - refreshable MVs are always refreshable."""
        return True

    def get_refresh_config(self) -> ResolvedRefreshConfig:
        """Returns the refresh configuration."""
        return self.refresh_config

File: packages/ts-moose-lib/src/dmv2/sdk/materializedView.ts
import { ClickHouseEngines } from "../../dataModels/types";
import { Sql, toStaticQuery } from "../../sqlHelpers";
import { OlapConfig, OlapTable } from "./olapTable";
import { View } from "./view";
import { LifeCycle } from "./lifeCycle";
import { IJsonSchemaCollection } from "typia";
import { Column } from "../../dataModels/dataModelTypes";
import { getMooseInternal, isClientOnlyMode } from "../internal";
import { getSourceFileFromStack } from "../utils/stackTrace";

/**
 * Helper function to format a table reference as `database`.`table` or just `table`
 */
function formatTableReference(table: OlapTable<any> | View): string {
  const database =
    table instanceof OlapTable ? table.config.database : undefined;
  if (database) {
    return `\`${database}\`.\`${table.name}\``;
  }
  return `\`${table.name}\``;
}

// ============================================================================
// Refresh Configuration Types for Refreshable Materialized Views
// ============================================================================

/**
 * Supported time units for refresh intervals.
 * Maps directly to ClickHouse interval units.
 */
export type TimeUnit =
  | "second"
  | "minute"
  | "hour"
  | "day"
  | "week"
  | "month"
  | "year";

/**
 * Refresh interval using EVERY mode - periodic refresh at fixed times.
 * Example: { type: "every", value: 1, unit: "hour" } => REFRESH EVERY 1 HOUR
 */
export interface RefreshIntervalEvery {
  type: "every";
  /** The numeric value of the interval */
  value: number;
  /** The time unit for the interval */
  unit: TimeUnit;
}

/**
 * Refresh interval using AFTER mode - refresh after interval since last refresh.
 * Example: { type: "after", value: 30, unit: "minute" } => REFRESH AFTER 30 MINUTE
 */
export interface RefreshIntervalAfter {
  type: "after";
  /** The numeric value of the interval */
  value: number;
  /** The time unit for the interval */
  unit: TimeUnit;
}

/**
 * Refresh interval specification - either EVERY or AFTER mode.
 */
export type RefreshInterval = RefreshIntervalEvery | RefreshIntervalAfter;

/**
 * A duration specified as value + unit.
 * Used for offset and randomize configurations.
 */
export interface Duration {
  /** The numeric value */
  value: number;
  /** The time unit */
  unit: TimeUnit;
}

/**
 * Configuration for refreshable (scheduled) materialized views.
 *
 * @example
 * ```typescript
 * const mv = new RefreshableMaterializedView<Stats>({
 *   materializedViewName: "hourly_stats_mv",
 *   selectStatement: "SELECT count(*) as cnt FROM events",
 *   selectTables: [eventsTable],
 *   targetTable: { name: "hourly_stats" },
 *   refreshConfig: {
 *     interval: { type: "every", value: 1, unit: "hour" },
 *     offset: { value: 5, unit: "minute" },
 *   },
 * });
 * ```
 */
export interface RefreshConfig {
  /** The refresh interval (EVERY or AFTER) */
  interval: RefreshInterval;
  /**
   * Optional offset from interval start.
   * NOTE: Only valid with REFRESH EVERY, not REFRESH AFTER.
   */
  offset?: Duration;
  /** Optional randomization window */
  randomize?: Duration;
  /**
   * Other refreshable MVs this one depends on (will wait for them to refresh first).
   * Only accepts RefreshableMaterializedView objects - ensures type safety.
   */
  dependsOn?: RefreshableMaterializedView<any>[];
  /** Use APPEND mode instead of full refresh */
  append?: boolean;
}

/**
 * Internal representation of RefreshConfig with dependsOn converted to string names.
 * @internal
 */
export interface ResolvedRefreshConfig {
  interval: RefreshInterval;
  offset?: Duration;
  randomize?: Duration;
  /** Names of other MVs this one depends on (already resolved to strings) */
  dependsOn?: string[];
  append?: boolean;
}

// ============================================================================
// Incremental MaterializedView Configuration and Class
// ============================================================================

/**
 * Configuration options for creating an incremental Materialized View.
 *
 * Incremental MVs are triggered on every insert to source tables.
 *
 * @template T The data type of the records stored in the target table.
 */
export interface MaterializedViewConfig<T> {
  /** The SQL SELECT statement or `Sql` object defining the data to be materialized. */
  selectStatement: string | Sql;

  /**
   * The source tables/views that the SELECT statement reads from.
   * These tables trigger the MV on insert.
   */
  selectTables: (OlapTable<any> | View)[];

  /** The name for the ClickHouse MATERIALIZED VIEW object itself. */
  materializedViewName: string;

  /**
   * The target table where transformed data is written.
   * Can be an existing OlapTable or inline configuration to create one.
   */
  targetTable?:
    | OlapTable<T>
    | {
        /** The name for the underlying target OlapTable. */
        name: string;
        /**
         * ClickHouse engine for the target table (e.g., ReplacingMergeTree).
         * Defaults to MergeTree.
         */
        engine?: ClickHouseEngines;
        /** Optional ordering fields for the target table. */
        orderByFields?: (keyof T & string)[];
      };

  /**
   * @deprecated Use targetTable instead.
   */
  tableName?: string;

  /**
   * @deprecated Use targetTable instead.
   */
  engine?: ClickHouseEngines;

  /**
   * @deprecated Use targetTable instead.
   */
  orderByFields?: (keyof T & string)[];

  /** Optional metadata for the materialized view. */
  metadata?: { [key: string]: any };

  /** Optional lifecycle management policy for the materialized view.
   * Controls whether Moose can drop or modify the MV automatically.
   * Defaults to FULLY_MANAGED if not specified. */
  lifeCycle?: LifeCycle;
}

/**
 * Represents an incremental Materialized View in ClickHouse.
 *
 * Incremental MVs are triggered on every insert to source tables.
 * For scheduled/refreshable MVs, use `RefreshableMaterializedView` instead.
 *
 * @template TargetTable The data type of the records stored in the target table.
 */
export class MaterializedView<TargetTable> {
  /** @internal */
  public readonly kind = "MaterializedView";

  /** The name of the materialized view */
  name: string;

  /** The target OlapTable instance where the materialized data is stored. */
  targetTable: OlapTable<TargetTable>;

  /** The SELECT SQL statement */
  selectSql: string;

  /** Names of source tables that the SELECT reads from */
  sourceTables: string[];

  /** Optional metadata for the materialized view */
  metadata: { [key: string]: any };

  /** Optional lifecycle management policy for the materialized view */
  lifeCycle?: LifeCycle;

  /**
   * Always undefined for incremental MVs.
   * @internal
   */
  refreshConfig?: undefined;

  /**
   * Creates a new incremental MaterializedView instance.
   *
   * @param options Configuration options for the materialized view.
   */
  constructor(options: MaterializedViewConfig<TargetTable>);

  /** @internal **/
  constructor(
    options: MaterializedViewConfig<TargetTable>,
    targetSchema: IJsonSchemaCollection.IV3_1,
    targetColumns: Column[],
  );
  constructor(
    options: MaterializedViewConfig<TargetTable>,
    targetSchema?: IJsonSchemaCollection.IV3_1,
    targetColumns?: Column[],
  ) {
    // Validate selectStatement
    let selectStatement = options.selectStatement;
    if (typeof selectStatement !== "string") {
      selectStatement = toStaticQuery(selectStatement);
    }

    if (targetSchema === undefined || targetColumns === undefined) {
      throw new Error(
        "Supply the type param T so that the schema is inserted by the compiler plugin.",
      );
    }

    // Validate selectTables is provided and not empty
    if (!options.selectTables || options.selectTables.length === 0) {
      throw new Error(
        "MaterializedView requires 'selectTables' to be specified with at least one table. " +
          "These are the tables that your SELECT statement reads from.",
      );
    }

    // Resolve target table configuration
    let targetTable: OlapTable<TargetTable>;

    if (options.targetTable !== undefined && options.tableName !== undefined) {
      throw new Error(
        "Cannot specify both 'targetTable' and 'tableName'. " +
          "Use 'targetTable' (preferred) or 'tableName' (deprecated).",
      );
    }

    if (options.targetTable instanceof OlapTable) {
      targetTable = options.targetTable;
    } else if (options.targetTable) {
      targetTable = new OlapTable(
        options.targetTable.name,
        {
          orderByFields: options.targetTable.orderByFields,
          engine: options.targetTable.engine ?? ClickHouseEngines.MergeTree,
        } as OlapConfig<TargetTable>,
        targetSchema,
        targetColumns,
      );
    } else if (options.tableName) {
      targetTable = new OlapTable(
        options.tableName,
        {
          orderByFields: options.orderByFields,
          engine: options.engine ?? ClickHouseEngines.MergeTree,
        } as OlapConfig<TargetTable>,
        targetSchema,
        targetColumns,
      );
    } else {
      throw new Error(
        "Target table must be specified. Use one of:\n" +
          "  targetTable: myOlapTable              // existing OlapTable\n" +
          "  targetTable: { name: 'my_table' }     // inline config\n" +
          "  tableName: 'my_table'                 // deprecated",
      );
    }

    if (targetTable.name === options.materializedViewName) {
      throw new Error(
        "Materialized view name cannot be the same as the target table name.",
      );
    }

    this.name = options.materializedViewName;
    this.targetTable = targetTable;
    this.selectSql = selectStatement;
    this.refreshConfig = undefined;

    // Set sourceTables from the selectTables configuration
    this.sourceTables = options.selectTables.map((t) =>
      formatTableReference(t),
    );
    this.lifeCycle = options.lifeCycle;

    // Initialize metadata
    this.metadata = options.metadata ? { ...options.metadata } : {};

    // Capture source file from stack trace if not already provided
    if (!this.metadata.source) {
      const stack = new Error().stack;
      const sourceInfo = getSourceFileFromStack(stack);
      if (sourceInfo) {
        this.metadata.source = { file: sourceInfo };
      }
    }

    // Register in the materializedViews registry
    const materializedViews = getMooseInternal().materializedViews;
    if (!isClientOnlyMode() && materializedViews.has(this.name)) {
      throw new Error(`MaterializedView with name ${this.name} already exists`);
    }
    materializedViews.set(this.name, this);
  }

  /**
   * Returns true - incremental MVs are always incremental.
   */
  isIncremental(): boolean {
    return true;
  }

  /**
   * Returns false - incremental MVs are never refreshable.
   */
  isRefreshable(): boolean {
    return false;
  }
}

// ============================================================================
// Refreshable MaterializedView Configuration and Class
// ============================================================================

/**
 * Configuration options for creating a Refreshable Materialized View.
 *
 * Refreshable MVs run on a schedule (REFRESH EVERY/AFTER) rather than
 * being triggered by inserts to source tables.
 *
 * Note: Refreshable MVs always use MergeTree engine (no custom engine option).
 *
 * @template T The data type of the records stored in the target table.
 */
export interface RefreshableMaterializedViewConfig<T> {
  /** The SQL SELECT statement or `Sql` object defining the data to be materialized. */
  selectStatement: string | Sql;

  /**
   * The source tables/views that the SELECT statement reads from.
   * These define the data lineage for the refresh operation.
   */
  selectTables: (OlapTable<any> | View)[];

  /** The name for the ClickHouse MATERIALIZED VIEW object itself. */
  materializedViewName: string;

  /**
   * The target table where transformed data is written.
   * Can be an existing OlapTable or inline configuration to create one.
   *
   * Note: Custom engines are not supported for refreshable MVs (MergeTree only).
   */
  targetTable:
    | OlapTable<T>
    | {
        /** The name for the underlying target OlapTable. */
        name: string;
        /** Optional ordering fields for the target table. */
        orderByFields?: (keyof T & string)[];
      };

  /**
   * Configuration for the refresh schedule.
   * Required for refreshable MVs.
   */
  refreshConfig: RefreshConfig;

  /** Optional metadata for the materialized view. */
  metadata?: { [key: string]: any };
}

/**
 * Represents a Refreshable Materialized View in ClickHouse.
 *
 * Refreshable MVs run on a schedule (REFRESH EVERY/AFTER) rather than
 * being triggered by inserts to source tables.
 *
 * For incremental/trigger-based MVs, use `MaterializedView` instead.
 *
 * @template TargetTable The data type of the records stored in the target table.
 */
export class RefreshableMaterializedView<TargetTable> {
  /** @internal */
  public readonly kind = "MaterializedView";

  /** The name of the materialized view */
  name: string;

  /** The target OlapTable instance where the materialized data is stored. */
  targetTable: OlapTable<TargetTable>;

  /** The SELECT SQL statement */
  selectSql: string;

  /** Names of source tables that the SELECT reads from */
  sourceTables: string[];

  /** Optional metadata for the materialized view */
  metadata: { [key: string]: any };

  /**
   * The refresh configuration for this refreshable MV.
   * dependsOn is resolved to string names.
   * @internal
   */
  refreshConfig: ResolvedRefreshConfig;

  /**
   * Creates a new RefreshableMaterializedView instance.
   *
   * @param options Configuration options for the materialized view.
   */
  constructor(options: RefreshableMaterializedViewConfig<TargetTable>);

  /** @internal **/
  constructor(
    options: RefreshableMaterializedViewConfig<TargetTable>,
    targetSchema: IJsonSchemaCollection.IV3_1,
    targetColumns: Column[],
  );
  constructor(
    options: RefreshableMaterializedViewConfig<TargetTable>,
    targetSchema?: IJsonSchemaCollection.IV3_1,
    targetColumns?: Column[],
  ) {
    // Validate selectStatement
    let selectStatement = options.selectStatement;
    if (typeof selectStatement !== "string") {
      selectStatement = toStaticQuery(selectStatement);
    }

    if (targetSchema === undefined || targetColumns === undefined) {
      throw new Error(
        "Supply the type param T so that the schema is inserted by the compiler plugin.",
      );
    }

    // Validate selectTables is provided and not empty
    if (!options.selectTables || options.selectTables.length === 0) {
      throw new Error(
        "RefreshableMaterializedView requires 'selectTables' to be specified with at least one table. " +
          "These are the tables that your SELECT statement reads from.",
      );
    }

    // Resolve target table configuration
    let targetTable: OlapTable<TargetTable>;

    if (options.targetTable instanceof OlapTable) {
      targetTable = options.targetTable;
      // Validate no custom engine on existing OlapTable
      const config = targetTable.config as { engine?: ClickHouseEngines };
      if (
        config.engine !== undefined &&
        config.engine !== ClickHouseEngines.MergeTree
      ) {
        throw new Error(
          "RefreshableMaterializedView cannot use custom engines. " +
            `Found engine '${config.engine}' but refreshable MVs only support MergeTree.`,
        );
      }
    } else {
      targetTable = new OlapTable(
        options.targetTable.name,
        {
          orderByFields: options.targetTable.orderByFields,
          engine: ClickHouseEngines.MergeTree, // Always MergeTree
        } as OlapConfig<TargetTable>,
        targetSchema,
        targetColumns,
      );
    }

    if (targetTable.name === options.materializedViewName) {
      throw new Error(
        "Materialized view name cannot be the same as the target table name.",
      );
    }

    // Validate interval value is a positive integer
    const intervalVal = options.refreshConfig.interval.value;
    if (!Number.isInteger(intervalVal) || intervalVal <= 0) {
      throw new Error(
        `Refresh interval value must be a positive integer, got: ${intervalVal}`,
      );
    }

    // Validate offset/randomize values if present
    for (const key of ["offset", "randomize"] as const) {
      const dur = options.refreshConfig[key];
      if (dur && (!Number.isInteger(dur.value) || dur.value <= 0)) {
        throw new Error(
          `Refresh ${key} value must be a positive integer, got: ${dur.value}`,
        );
      }
    }

    // Validate OFFSET is not used with REFRESH AFTER (only valid with REFRESH EVERY)
    if (
      options.refreshConfig.interval.type === "after" &&
      options.refreshConfig.offset
    ) {
      throw new Error(
        "OFFSET is only valid with REFRESH EVERY, not REFRESH AFTER. " +
          "Remove the 'offset' option or change the interval type to 'every'.",
      );
    }

    this.name = options.materializedViewName;
    this.targetTable = targetTable;
    this.selectSql = selectStatement;

    // Convert refreshConfig.dependsOn from RefreshableMaterializedView objects to string names
    this.refreshConfig = {
      ...options.refreshConfig,
      dependsOn: options.refreshConfig.dependsOn?.map((dep) => dep.name),
    };

    // Set sourceTables from the selectTables configuration
    this.sourceTables = options.selectTables.map((t) =>
      formatTableReference(t),
    );

    // Initialize metadata
    this.metadata = options.metadata ? { ...options.metadata } : {};

    // Capture source file from stack trace if not already provided
    if (!this.metadata.source) {
      const stack = new Error().stack;
      const sourceInfo = getSourceFileFromStack(stack);
      if (sourceInfo) {
        this.metadata.source = { file: sourceInfo };
      }
    }

    // Register in the materializedViews registry
    const materializedViews = getMooseInternal().materializedViews;
    if (!isClientOnlyMode() && materializedViews.has(this.name)) {
      throw new Error(
        `A MaterializedView with name '${this.name}' is already registered.`,
      );
    }
    materializedViews.set(this.name, this);
  }

  /**
   * Returns false - refreshable MVs are never incremental.
   */
  isIncremental(): boolean {
    return false;
  }

  /**
   * Returns true - refreshable MVs are always refreshable.
   */
  isRefreshable(): boolean {
    return true;
  }

  /**
   * Returns the refresh configuration.
   */
  getRefreshConfig(): ResolvedRefreshConfig {
    return this.refreshConfig;
  }
}
